sw-emulator/lib/crypto/src/sha512.rs

```diff
@@ -118,14 +126,60 @@ impl Sha512 {
     ///
     /// # Arguments
     ///
-    /// * `block` - Block to compress
+    /// * `block` - Block to compress. Each word in `block` is expected to
+    ///   be little-endian
     pub fn update(&mut self, block: &[u8; Self::BLOCK_SIZE]) {
         let mut block = *Block::<U128>::from_slice(block);
+        block.to_big_endian();
 
-        // Block is received as a list of big-endian DWORDs.
-        // Changing them to little-endian.
-        block.to_little_endian();
         sha2::compress512(&mut self.hash, &[block]);
+        self.blocks_processed += 1;
+    }
+
+    /// Update the hash with an arbitrary number to bytes
+    ///
+    /// `bytes` is expected to be big-endian data
+    pub fn update_bytes(&mut self, bytes: &[u8]) {
+        self.partial_block.extend_from_slice(bytes);
+
+        while self.partial_block.len() >= Self::BLOCK_SIZE {
+            // Safe to unwrap becasue slice is guaranteed to be correct size
```

Comment:
Nit: it's possible to use https://doc.rust-lang.org/std/primitive.slice.html#method.get_mut here to get rid of the unwrap

Reply:
The borrow checker doesn't like

```rust
 while let Some(block) = self.partial_block.get_mut(..Self::BLOCK_SIZE) {
    self.update(block.try_into().unwrap());
}
```

Because we have a mutable reference to `self` (from get_mut) and `self.update` also takes a mutable reference to `self`. I'm not sure if there's a better way to do this.

Reply:
Yeah, the borrow checker doesn't seem to offer any mechanism to demote the &mut reference to a &.